Web: Refactor how we use toast notifications

kimlisa opened the pull request:
recommend reviewing by commit

refactoring toast notification as a result of comments in https://github.com/gravitational/teleport/pull/56534

This PR adds a global toast notification (`Main.tsx`) through context provider and replaces and removes repetitive notification logic.

The toast notification will render top right, and is sticky when scrolling.

Tested the following toast rendering for (i also checked the stories related renders):
- account (adding/removing passkey)
- roles (create/edit)
- user add (only in cloud)
- invite collab (this one i tested by mocking data)
- access list reviews

demo sticky floating note:
https://github.com/user-attachments/assets/3a913fe8-e976-4bc3-b419-af82780927c6


with top bar alert and a info guide rendered:
<img width="1728" height="411" alt="image" src="https://github.com/user-attachments/assets/45d13e98-0c54-4e02-b4aa-11d5ca8649ce" />

without banners and guides:
<img width="1728" height="340" alt="image" src="https://github.com/user-attachments/assets/da2ec8c1-e33b-408c-aae2-d532a4e039f5" />

multi toasts (i changed the windows keyboard multi times in a row, it says the same thing)
<img width="1728" height="441" alt="image" src="https://github.com/user-attachments/assets/92e8c6a4-ceb3-40f3-ae0c-379f04cad793" />

kimlisa committed:
Rename shared/Notification to ToastNotification

kimlisa committed:
Implement toast notification context and provider

kimlisa committed:
Add toast context provider

kimlisa committed:
Replace old notification with the new one

kimlisa committed:
Update fields for InviteCollaborators

kimlisa commented on web/packages/teleport/src/Teleport.tsx:
```diff
@@ -101,15 +103,17 @@ const Teleport: React.FC<Props> = props => {
-                            <Route>{createPrivateRoutes()}</Route>
-                          </Switch>
-                        </TeleportContextProvider>
+                        <ToastNotificationProvider>
+                          <TeleportContextProvider ctx={ctx}>
+                            <Switch>
+                              <Route
+                                path={cfg.routes.appLauncher}
+                                component={AppLauncher}
+                              />
+                              <Route>{createPrivateRoutes()}</Route>
+                            </Switch>
+                          </TeleportContextProvider>
+                        </ToastNotificationProvider>
```
i had to put the toast provider here because of how we handle user invites for starting up cloud cluster. changes found here: https://github.com/gravitational/teleport.e/pull/6908

This is the flow:
- start up cloud cluster
- create user/password
- before going to the dashboard `Main.tsx`, it asks you to invite collaborators
- afterwards, the `Main.tsx` renders, and then we render the toast notifying users if invitation succeeded or not

bl-nero replied:
Doesn't it mean that we'll have two nested toast notification providers, and depending on route, we'll be using one or another?

kimlisa replied:
hrm, i wasn't entirely sure what you meant. there is only one `ToastNotificationProvider` defined atm 🤔

bl-nero replied:
Oops, I must have either confused `TestNotifications` with `TestNotificationProvider` or confused fragments of `Main.tsx` with `Main.test.tsx`. Never mind. :)

bl-nero reviewed (commented):
Finally someone decided to refactor this. Much needed. Thank you, Lisa!

kimlisa commented on web/packages/teleport/src/Main/Main.test.tsx:
```diff
@@ -120,76 +133,90 @@ test('toggle rendering of info guide panel', async () => {
 });
 
-test('displays invite collaborators feedback if present', () => {
+test('notification render and auto dismissal', async () => {
```
the git diff failed on this one, i deleted these tests as the [invite collaborators field](https://github.com/gravitational/teleport/blob/67bcca9f15f30dfe1b79e39ca22342a74e8f8331/web/packages/teleport/src/Main/Main.tsx#L70) got deleted so the test wasn't relevant anymore. 

the `inviteCollaboratorsFeedback` existed as a hack to render toast notifications from enterprise, but since we have a provider now, this hack isn't required anymore.

kimlisa committed:
Add comments, switch ordering of notes

kimlisa commented on web/packages/shared/components/ToastNotification/ToastNotificationContext.tsx:
```diff
@@ -59,8 +69,8 @@ export const ToastNotificationProvider: FC<PropsWithChildren> = ({
     content: ToastNotificationItemContent
   ) {
     setNotifications(notifications => [
-      ...notifications,
```
i made it so that the newest notes gets added to the beginning of list so when it renders, the order of toasts from top of screen is `newest -> oldest`

bl-nero commented on web/packages/shared/components/ToastNotification/ToastNotifications.tsx:
```diff
@@ -0,0 +1,65 @@
+  );
+};
+
+const TopRightStickyContainer = styled.div`
+  position: sticky;
+  top: 0;
+  right: 0;
+  z-index: 1;
+`;
+
+const TopRightFlexedContainer = styled.div`
+  position: absolute;
+  top: ${p => p.theme.space[2]}px;
+  right: ${p => p.theme.space[5]}px;
+  display: flex;
+  flex-direction: column;
+  gap: ${p => p.theme.space[3]}px;
+`;
```
There's one small issue about putting it in the top right corner: our CTA buttons are also usually there, so when someone wants to repeat the action, they'd have to dismiss the notification first. However, I don't really know if it's worth moving these to the bottom, as they will be less prominent there.

kimlisa committed:
Address CR

When adding new notes, pass in object instead of multi params
Remove duplicate in the name eg: addNotification -> add

kimlisa commented on web/packages/shared/components/ToastNotification/ToastNotificationContext.tsx:
```diff
@@ -0,0 +1,109 @@
+    []
+  );
+
+  function remove(id: string) {
+    setNotifications(n => n.filter(item => item.id !== id));
+  }
+
+  function add(entry: NotificationEntry) {
+    setNotifications(notifications => [
+      {
+        id: crypto.randomUUID(),
+        content: entry.content,
+        severity: entry.severity,
+      },
+      ...notifications,
+    ]);
+  }
```
i did wrap the functions in [useMemo here](https://github.com/gravitational/teleport/pull/56832/files#diff-6753c64cc6cd288edc4e036dd8feb5412213e8b5c5b5bb8f31e4782bc70134e0R80) (similar pattern i got from somewhere else), or do you prefer individually wrapped callbacks?

kimlisa replied:
how about this: https://github.com/gravitational/teleport/pull/56832/commits/ad3b3b4ddb6617fb9c9e9b0558f0e313ddee8445

rudream reviewed (approved):
LGTM once @ryanclark comment is addressed, we should be memoizing the context values to prevent unnecessary re-renders

kimlisa committed:
Wrap context funcs in useCallbacks
